### PR TITLE
Allow output of network pressures based on rates at end of time step

### PR DESCRIPTION
**Pull request by vkip**

### CODE REVIEW

**Comment by vkip** [comment]
jenkins build this opm-common=3701 please

**Comment by GitPaean** [comment]
In this PR, it looks like it will generate the SUMMARY output with GPR:WELLNAME.  And in https://github.com/OPM/opm-common/pull/3701, it is mentioned that NPR and GNETPR are introduced. 

Please confirm you indeed want `GPR:WELLNAME` SUMMARY output. @vkip

**Comment by vkip** [comment]
> In this PR, it looks like it will generate the SUMMARY output with GPR:WELLNAME. And in [OPM/opm-common#3701](https://github.com/OPM/opm-common/pull/3701), it is mentioned that NPR and GNETPR are introduced.
> 
> Please confirm you indeed want `GPR:WELLNAME` SUMMARY output. @vkip

Should have explicitly mentioned that in the description, but yes, I would also like GPR:WELLNAME output (in addition to NPR/GNETPR:WELLNAME) for convenience of comparing the pressures coming from the network without looking up the group.

**Review by bska — COMMENTED** [review]
I think I see the same kind of general problem here as I noticed in the upstream PR.  We may need to be a little more careful about treating nodes as groups, because there is not necessarily a direct translation between those in runs using the extended network model.

File: opm/simulators/wells/BlackoilWellModelGeneric.cpp [hunk 1]
```diff
@@ -902,11 +902,40 @@ assignGroupValues(const int                               reportStepIdx,
...
-assignNodeValues(std::map<std::string, data::NodeData>& nodevalues) const
+assignNodeValues(std::map<std::string, data::NodeData>& nodevalues, const int reportStepIdx) const
 {
     nodevalues.clear();
+    if (reportStepIdx < 0) return;
+
     for (const auto& [node, pressure] : node_pressures_) {
         nodevalues.emplace(node, data::NodeData{pressure});
+        // Assign node values of group to GPR:WELLNAME
+        const auto& group = schedule().getGroup(node, reportStepIdx);
```
[the remark this answers is not in the document]
Comment (vkip) [reply opm/simulators/wells/BlackoilWellModelGeneric.cpp:913]: Thanks! Adding the hasGroup() check.

File: opm/simulators/wells/BlackoilWellModelGeneric.cpp [hunk 2]
```diff
@@ -902,11 +902,40 @@ assignGroupValues(const int                               reportStepIdx,
 
 void
 BlackoilWellModelGeneric::
-assignNodeValues(std::map<std::string, data::NodeData>& nodevalues) const
+assignNodeValues(std::map<std::string, data::NodeData>& nodevalues, const int reportStepIdx) const
 {
     nodevalues.clear();
+    if (reportStepIdx < 0) return;
```
[the remark this answers is not in the document]
Comment (vkip) [reply opm/simulators/wells/BlackoilWellModelGeneric.cpp:908]: Always happy to remove unnecessary lines!

**Comment by GitPaean** [comment]
jenkins build this opm-common=3701 please

**Review by bska — COMMENTED** [review]
Having pondered this a bit more, it's not clear to me that the well-name based quantity, at least as currently implemented, is well defined in a run with multi-tiered network (extended model) or group tree (standard network).  Unless we can expect every node along one branch of the network/group tree to have the **same** converged network pressure we risk assigning a well value that's more or less random&ndash;or at least based on the visiting order of the network nodes or the group tree.

If we want to define the pressure tagged with a well name more strictly&ndash;e.g., the network pressure of the node that is closest to the well&ndash;then that is well defined.  That definition will, at least as far as I can tell, require different logic than what is implemented here, however.

File: opm/simulators/wells/BlackoilWellModelGeneric.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (vkip) [reply opm/simulators/wells/BlackoilWellModelGeneric.cpp]: ... but turns the check is in fact required - the return statement in the [external logic](https://github.com/OPM/opm-simulators/blob/86c919877edd7fd9952182c9c6a216546b3545be/ebos/eclwriter.hh#L176-L195) seems never to be reached (unless you start counting at -1..? )

**Comment by vkip** [comment]
> Having pondered this a bit more, it's not clear to me that the well-name based quantity, at least as currently implemented, is well defined in a run with multi-tiered network (extended model) or group tree (standard network). Unless we can expect every node along one branch of the network/group tree to have the **same** converged network pressure we risk assigning a well value that's more or less random–or at least based on the visiting order of the network nodes or the group tree.
> 
> If we want to define the pressure tagged with a well name more strictly–e.g., the network pressure of the node that is closest to the well–then that is well defined. That definition will, at least as far as I can tell, require different logic than what is implemented here, however.

Not following you here. Groups cannot contain both wells and other groups, and as far as I can tell the call to 'group.wells()' will only return wells actually contained in the group itself (not all wells found down the subtree). So the pressure of the unique node/group to which the well belongs should be a well defined quantity? (I realize I'm probably missing something, so please feel free to spell it out in embarrassing detail. )

**Comment by vkip** [comment]
jenkins build this opm-common=3701 please

**Comment by bska** [comment]
jenkins build this please

**Review by bska — APPROVED** [review]
Thanks a lot for the updates.  The new reference solutions have been installed on the CI system so I'll merge this to activate the support.

**Pull request merged**